Update readme.md (#193)
eclipse/kiso: examples/CommonGateway/c-sensors/readme.md

# Example Project: C-sensor for CommonGateway Development Kit

## Goal:
This example is for running the c-sensor in CommonGateway.

## Brief:
This example will show the enviromental measurements and dimensions.

### Components Required:

#### Software Components
Please refer to the getting started [quick start](http://docs.eclipsekiso.de:1313/user-guide/quick_start.html) guide for Eclipse Kiso.

#### Hardware Components
You will need CommonGateway Board. 
## Initial Setup

> Configure the project to be compiled for the CommonGateway Board which is located in [boards/CommonGateway](https://github.com/eclipse/kiso/tree/master/boards/CommonGateway) .

Reviewer: ```suggestion
Configure the project to be compiled for the CommonGateway Board which is located in [boards/CommonGateway](https://github.com/eclipse/kiso/tree/master/boards/CommonGateway). To do this, add the respective `KISO_BOARD_NAME` setting to `cmake` during workspace configuration, as shown in Section [Execution Flow](#Execution-Flow)
```

## Execution Flow
```
cd kiso
cmake . -Bbuild -DKISO_BOARD_NAME=CommonGateway -DKISO_APPLICATION_NAME=c-sensors -DKISO_APPLICATION_PATH=examples/CommonGateway/c-sensors 
cmake --build build
```
Now as you have built the project you should have an application binary ready to run.
For you own convenience you can use the cmake build system also to flash the binary with the external debugger.
```
cmake --build build --target flash
```

## Code Walkthrough

> The essence of this example can be seen in the appPeriodicTest function, which is taking the data from  SensorEnvironment_Read and make some calculations to give the output understandable and take the data from the SensorAccelerometer_Read and show dimensions and if everything goes correct, the green led will flash. Otherwise the red led will flash .

Reviewer: ```suggestion
The main program loop can be seen in `appPeriodicTest()`. It takes data from  `SensorEnvironment_Read()` and `SensorAccelerometer_Read()`, performs some conversions and corrections based on static calibration and finally presents the data via debug log. If everything went OK, the green LED will flash. In case of any errors the red LED will flash.
```

```c
void appPeriodicTest(void *CmdProcessorHandle, uint32_t param2)
{
    (void)param2;
    struct bme280_data sensorEnvironmentData = {0};
    struct bma2x2_accel_data sensorAccelerometerData = {0};

    Retcode_T returnValue = RETCODE_OK;

    if (RETCODE_OK == returnValue)
    {
        returnValue = SensorEnvironment_Read(&sensorEnvironmentData);
        if (RETCODE_OK == returnValue)
        {
#ifdef BME280_FLOAT_ENABLE
#error BME280_FLOAT_ENABLE not supported for this example
#else
            /* Finally we need to apply static temperature offset correction.
             * As no two sensors can be the same, this coefficient will be
             * different from sensor to sensor.
             * */
            const int32_t DEFAULT_TEMPERATURE_OFFSET = -210;
            int32_t temperature = sensorEnvironmentData.temperature + DEFAULT_TEMPERATURE_OFFSET;

>             //  Unit is in 1024 * % relative humidity
>             uint32_t humidity = (sensorEnvironmentData.humidity * 1000) / 1024;
> #ifdef BME280_64BIT_ENABLE
>             // If macro "BME280_64BIT_ENABLE" is enabled, which it is by default, the unit is 100 * Pascal
>             uint32_t pressure = sensorEnvironmentData.pressure;
> #else
>             // If this macro is disabled, Then the unit is in Pascal
>             uint32_t pressure = sensorEnvironmentData.pressure * 100;
> #endif
>             // clang-format off
>             LOG_DEBUG("Temperature: %ld.%02lu;\tPresure: %lu.%03luhPa;\tHumidity: %lu.%03lu%%RH",
>                     temperature / 100,
>                     temperature % 100,
>                     pressure / 10000,
>                     (pressure % 10000) / 10,
>                     humidity / 1000,
>                     humidity % 1000);
>             // clang-format on
> #endif
>         }
>         else
>         {
>             LOG_DEBUG("Env sensor read failed!\r\n");
>         }
>     }
> 
>     if (RETCODE_OK == returnValue)
>     {
>         returnValue = SensorAccelerometer_Read(&sensorAccelerometerData);
>         if (RETCODE_OK == returnValue)
>         {
>             // clang-format off
>             LOG_DEBUG("X: %ld;\tY: %ld;\tZ: %ld",
>                     sensorAccelerometerData.x,
>                     sensorAccelerometerData.y,
>                     sensorAccelerometerData.z);
>             // clang-format on
>         }
>         else
>         {
>             LOG_DEBUG("Accel sensor read failed!\r\n");
>         }
>     }
> 
>     if (RETCODE_OK == returnValue)
>     {
>         BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_TOGGLE);
>     }
>     else
>     {
>         BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_OFF);
>         BSP_LED_Switch(COMMONGATEWAY_LED_RED_ID, COMMONGATEWAY_LED_COMMAND_ON);
>     }
> 
>     if (RETCODE_OK == returnValue)
>     {
>         BSP_Board_Delay(5);
>         returnValue = CmdProcessor_Enqueue((CmdProcessor_T *)CmdProcessorHandle, appPeriodicTest, CmdProcessorHandle, UINT32_C(0));
>     }
> }

Reviewer: ```suggestion
void appPeriodicTest(void *CmdProcessorHandle, uint32_t param2)
{
    struct bme280_data sensorEnvironmentData = {0};
    struct bma2x2_accel_data sensorAccelerometerData = {0};

    Retcode_T returnValue = RETCODE_OK;

    // [...]

    if (RETCODE_OK == returnValue)
    {
        returnValue = SensorEnvironment_Read(&sensorEnvironmentData);
    
        // [... perform conversion and corrections based on calibration coefficients ...]
    
        if (RETCODE_OK == returnValue)
        {
            LOG_DEBUG("Temperature: %ld.%02lu;\tPresure: %lu.%03luhPa;\tHumidity: %lu.%03lu%%RH",
                    temperature / 100,
                    temperature % 100,
                    pressure / 10000,
                    (pressure % 10000) / 10,
                    humidity / 1000,
                    humidity % 1000);
        }
        else
        {
            LOG_DEBUG("Env sensor read failed!\r\n");
        }
    }

    if (RETCODE_OK == returnValue)
    {
        returnValue = SensorAccelerometer_Read(&sensorAccelerometerData);
        
        if (RETCODE_OK == returnValue)
        {
            LOG_DEBUG("X: %ld;\tY: %ld;\tZ: %ld",
                    sensorAccelerometerData.x,
                    sensorAccelerometerData.y,
                    sensorAccelerometerData.z);
        }
        else
        {
            LOG_DEBUG("Accel sensor read failed!\r\n");
        }
    }

    if (RETCODE_OK == returnValue)
    {
        BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_TOGGLE);
    }
    else
    {
        BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_OFF);
        BSP_LED_Switch(COMMONGATEWAY_LED_RED_ID, COMMONGATEWAY_LED_COMMAND_ON);
    }

    if (RETCODE_OK == returnValue)
    {
        BSP_Board_Delay(5);
        returnValue = CmdProcessor_Enqueue((CmdProcessor_T *)CmdProcessorHandle, appPeriodicTest, CmdProcessorHandle, UINT32_C(0));
    }

    // [...]
}
```

>             //  Unit is in 1024 * % relative humidity
>             uint32_t humidity = (sensorEnvironmentData.humidity * 1000) / 1024;
> #ifdef BME280_64BIT_ENABLE
>             // If macro "BME280_64BIT_ENABLE" is enabled, which it is by default, the unit is 100 * Pascal
>             uint32_t pressure = sensorEnvironmentData.pressure;
> #else
>             // If this macro is disabled, Then the unit is in Pascal
>             uint32_t pressure = sensorEnvironmentData.pressure * 100;
> #endif
>             // clang-format off
>             LOG_DEBUG("Temperature: %ld.%02lu;\tPresure: %lu.%03luhPa;\tHumidity: %lu.%03lu%%RH",
>                     temperature / 100,
>                     temperature % 100,
>                     pressure / 10000,
>                     (pressure % 10000) / 10,
>                     humidity / 1000,
>                     humidity % 1000);
>             // clang-format on
> #endif
>         }
>         else
>         {
>             LOG_DEBUG("Env sensor read failed!\r\n");
>         }
>     }
> 
>     if (RETCODE_OK == returnValue)
>     {
>         returnValue = SensorAccelerometer_Read(&sensorAccelerometerData);
>         if (RETCODE_OK == returnValue)
>         {
>             // clang-format off
>             LOG_DEBUG("X: %ld;\tY: %ld;\tZ: %ld",
>                     sensorAccelerometerData.x,
>                     sensorAccelerometerData.y,
>                     sensorAccelerometerData.z);
>             // clang-format on
>         }
>         else
>         {
>             LOG_DEBUG("Accel sensor read failed!\r\n");
>         }
>     }
> 
>     if (RETCODE_OK == returnValue)
>     {
>         BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_TOGGLE);
>     }
>     else
>     {
>         BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_OFF);
>         BSP_LED_Switch(COMMONGATEWAY_LED_RED_ID, COMMONGATEWAY_LED_COMMAND_ON);
>     }
> 
>     if (RETCODE_OK == returnValue)
>     {
>         BSP_Board_Delay(5);
>         returnValue = CmdProcessor_Enqueue((CmdProcessor_T *)CmdProcessorHandle, appPeriodicTest, CmdProcessorHandle, UINT32_C(0));
>     }
> }

Reviewer: ```suggestion
void appPeriodicTest(void *CmdProcessorHandle, uint32_t param2)
{
    struct bme280_data sensorEnvironmentData = {0};
    struct bma2x2_accel_data sensorAccelerometerData = {0};

    Retcode_T returnValue = RETCODE_OK;

    // [...]

    if (RETCODE_OK == returnValue)
    {
        returnValue = SensorEnvironment_Read(&sensorEnvironmentData);
    
        // [... perform conversion and corrections based on calibration coefficients ...]
    
        if (RETCODE_OK == returnValue)
        {
            LOG_DEBUG("Temperature: %ld.%02lu;\tPresure: %lu.%03luhPa;\tHumidity: %lu.%03lu%%RH",
                    temperature / 100,
                    temperature % 100,
                    pressure / 10000,
                    (pressure % 10000) / 10,
                    humidity / 1000,
                    humidity % 1000);
        }
        else
        {
            LOG_DEBUG("Env sensor read failed!\r\n");
        }
    }

    if (RETCODE_OK == returnValue)
    {
        returnValue = SensorAccelerometer_Read(&sensorAccelerometerData);
        
        if (RETCODE_OK == returnValue)
        {
            LOG_DEBUG("X: %ld;\tY: %ld;\tZ: %ld",
                    sensorAccelerometerData.x,
                    sensorAccelerometerData.y,
                    sensorAccelerometerData.z);
        }
        else
        {
            LOG_DEBUG("Accel sensor read failed!\r\n");
        }
    }

    if (RETCODE_OK == returnValue)
    {
        BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_TOGGLE);
    }
    else
    {
        BSP_LED_Switch(COMMONGATEWAY_LED_GREEN_ID, COMMONGATEWAY_LED_COMMAND_OFF);
        BSP_LED_Switch(COMMONGATEWAY_LED_RED_ID, COMMONGATEWAY_LED_COMMAND_ON);
    }

    if (RETCODE_OK == returnValue)
    {
        BSP_Board_Delay(5);
        returnValue = CmdProcessor_Enqueue((CmdProcessor_T *)CmdProcessorHandle, appPeriodicTest, CmdProcessorHandle, UINT32_C(0));
    }

    // [...]
}
```

```

> This code through in two main functions to get the output.
> The first one is for the Enviroment Sensor, which go through two functions called [SensorEnvironment_Init](https://github.com/eclipse/kiso/blob/master/examples/CommonGateway/c-sensors/source/env_sensor.c#L58) and [SensorEnvironment_Read](https://github.com/eclipse/kiso/blob/master/examples/CommonGateway/c-sensors/source/env_sensor.c#L110) as you can see that these two functions give the measurements continuously for the (temperature, pressure and humidity).
> The second one is for Accelerometer Sensor, which go through two functions called [SensorAccelerometer_Init](https://github.com/eclipse/kiso/blob/master/examples/CommonGateway/c-sensors/source/accel_sensor.c#L48) and [SensorAccelerometer_Read](https://github.com/eclipse/kiso/blob/master/examples/CommonGateway/c-sensors/source/accel_sensor.c#L76) as you can see that these two functions give the dimensions (x,y,z).

Reviewer: ```suggestion
The app is structured in three modules:

* the main app logic
* environment sensor and
* acceleration sensor logic.

With the main app logic being at top of the hierarchy.

Each module has to be initialized first, before it can be used. The app logic initializes vital system components such as the sensor BSP components, as well as the I2C communication bus they are sit on. Finally each sensor module initializes the `bma2x2.h` and `bme280.h` library modules respectively.

Once initialization is complete, the main application loop is entered. This is done using the main command processor, created in `main()`. For each iteration of the app loop, `appPeriodicTest()` is enqueued at the end of the last iteration. This is repeated endlessly.

Reading acceleration data from `SensorAccelerometer_Read()`, one can directly access the x, y and z components as provided by the sensor. Environmental data however, need some static temperature correction first. This compensation can be changed via the `DEFAULT_TEMPERATURE_OFFSET` constant, with the measured temperature offset at a known temperature.
```

## Expected Output

> Specify what the user should obtain as an output of the execution the example. e.g. terminal output screenshots, pictures(flashing LEDs) etc.

Reviewer: ```suggestion
During normal operation the green LED should toggle for each app loop cycle every 5 seconds. If an error occurred anywhere within the app loop iteration, the red LED will be set ON and the green LED set to OFF. The app loop cycle continues, and may recover, toggling the green LED, while keeping the red LED on.

At the same time, with each app loop, data is read from the sensors. This data can be viewed via the debug UART port, provided a appropriate host is connected.
```

## Troubleshooting

> If you don't find a conclusion using the [documentation](http://docs.eclipsekiso.de:1313/), please message us on [mattermost](mattermost.eclipse.org/eclipse/channels/kiso).

Reviewer: ```suggestion
If you don't find a conclusion using the [documentation](http://docs.eclipsekiso.de:1313/), please message us on [mattermost](https://mattermost.eclipse.org/eclipse/channels/kiso).
```
